### ISSUE TITLE
Excon::Errors::SocketError Broken pipe (Errno::EPIPE) when use Activeadmin upload image by nested form 

### ISSUE THREAD

**Issue by jianbo** [fog/fog#3597]
I have following association, this error only appears if I set storage: fog, 

    class Article < ActiveRecord::Base
	has_many :article_images
        accepts_nested_attributes_for :article_images, :allow_destroy => true
  end

class ArticleImage < ActiveRecord::Base
  belongs_to :article

  mount_uploader :image, ArticleImageUploader
end
    

And my uploader:

# encoding: utf-8

class ArticleImageUploader < CarrierWave::Uploader::Base
  include ::CarrierWave::Backgrounder::Delay
  # Include RMagick or MiniMagick support:
  # include CarrierWave::RMagick
  include CarrierWave::MiniMagick

  # Choose what kind of storage to use for this uploader:
  # storage :file
  storage :fog

  # Override the directory where uploaded files will be stored.
  # This is a sensible default for uploaders that are meant to be mounted:
  def store_dir
    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  end

  version :mobile_large do
    process :resize_to_fit => [430, 430]
  end
end

Nested form

 form do |f|
    f.actions
    f.inputs 'Countries / Regions' do
      f.input :tags, :as => :select, :input_html => {:multiple => true}
      f.input :title
      f.input :article_category_id, :label => 'Category', :as => :select, :collection => ArticleCategory.all.map{|u| [u.name, u.id]}
      f.input :article_level_id, :label => 'Level', :as => :select, :collection => ArticleLevel.all.map{|u| [u.name, u.id]}
      f.input :source_url
      f.input :content
      f.input :raw_content
      f.input :author
      f.input :article_created_at, :as => :datepicker

      f.has_many :article_images do |ff|
        ff.input :image
      end

    end
    f.actions
  end

**Comment by geemus** [comment]
I suspect this may be more of a carrierwave issue than fog or excon per se, but I'm not sure. Could you include more information about the backtrace you are getting when things go awry? Thanks!

**Comment by plribeiro3000** [comment]
@jianbo Could you give us more information about the issue? Thanks!

**Comment by fcheung** [comment]
I've run into this and figured out the cause (at least in my case)

If there is any sort of authorisation problem, then AWS will send back a 403 response, before excon has finished sending all the data & then kills the connection at some point. Excon sees the dropped connection (hence the exception) but not the 403 response.

This sounds like something that needs to be fixed in excon, but not quite sure how

**Comment by geemus** [comment]
Thanks for the extra details, I'm also not sure how best to handle that particular case. I'll give it some thought though, hopefully there can be something at least a bit more graceful we can do.

**Comment by geemus** [comment]
@fcheung I don't suppose you know if there are any specifics around when this does or does not happen? I suppose if you knew cases that might do this you could do 100-continue stuff or otherwise guard with permission checks. Otherwise, it seems like a pretty big change to try and read from the socket before the request has finished writing (right now I believe it writes everything, then reads everything). What do you think?

**Comment by fcheung** [comment]
This was a PUT request to create a file on S3. Turning on 100-continues for this one action (+ I suppose the corresponding multipart action) should be fairly simple and sounds like it would address the issue.

RFC2616 says that sending the response before the request is fully read is not allowed, but you can see why they did it.

**Comment by geemus** [comment]
Yeah. I guess it all depends on how often this happens. If it is common, doing 100-continue seems like a good way to guard. Agreed that they probably shouldn't do this, but that it makes sense. Not much sense it letting a client send an arbitrarily large amount of data and claim a socket if it all is getting thrown away in the end.

As long as this only happens in fairly constrained contexts, that may be good-enough. I haven't seen this often in any event. We may want to at least document this somewhere/somehow, if we don't have a clear way to easily deal with it programmatically.

**Comment by fcheung** [comment]
I think this only happens if the upload is denied with a 403. Amazon do seem to say that using 100-continue is a good thing, moreover given that S3 supports it (and any http/1.1 compliant server should) then it seems pretty harmless. Does excon already support 100-continue?

**Comment by geemus** [comment]
Not yet, I don't think anyway. Would have to think a bit about how best to integrate it, but it doesn't sound too complex anyway (details: http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3)
